docs/docs/reconciliation.md

```diff
@@ -0,0 +1,147 @@
+---
+id: reconciliation
+title: Reconciliation
+permalink: docs/reconciliation.html
+---
+
+React provides a declarative API so that you don't have to worry about exactly what changes on every update. This makes writing applications a lot easier, but it might not be obvious how this is implemented within React. This article explains the choices we made in React's "diffing" algorithm so that component updates are predictable while being fast enough for high-performance apps.
+
+## Motivation
+
+When you use React, at a single point in time you can think of the `render()` function as creating a tree of React elements. On the next state or props update, that `render()` function will return a different tree of React elements. React then needs to figure out a way to efficiently convert the first tree into the second tree.
+
+There are some generic solutions to this algorithmic problem of generating the minimum number of operations to transform one tree into another. However, the [state of the art algorithms](http://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) have a complexity in the order of O(n<sup>3</sup>) where n is the number of elements in the tree.
+
+If we used this in React, displaying 1000 elements would require in the order of one billion comparisons. This is far too expensive. Instead, React implements a heuristic O(n) algorithm based on two assumptions:
+
+1. Two elements of different types can be treated as totally different trees.
+2. The developer can hint at which child elements may be stable across different renders with a `key` prop.
+
+In practice, these assumptions are valid for almost all practical use cases.
+
+## The Diffing Algorithm
+
+When diffing two trees, React first compares the the two root elements. The behavior is different depending on the types of the root elements.
+
+### Elements Of Different Types
+
+Whenever the root elements have different types, React will tear down the old tree and build the new tree from scratch.
+
+When tearing down a tree, old DOM elements are destroyed. Component instances receive `componentWillUnmount()`. When building up a new tree, new DOM elements are inserted into the DOM. Component instances receive `componentWillMount()` and then `componentDidMount()`. Any state associated with the old tree is lost.
```

Comment:
Proposal: let's say "DOM node" in this document when we refer to browser DOM node. This will help remove some ambiguity with the "element" word which we'll use for React elements.